packages/sockets/src/logic/creators/createSocketCommunicator.ts

```diff
@@ -421,6 +421,47 @@ export default function createSocketCommunicator({
     }
   }
 
+  async function gatherv(
+    data: unknown[],
+    sendCount: number,
+    buf: unknown[],
+    recvCounts: number[],
+    recvOffsets: number[],
+    root: number,
+    abort?: Abort,
+  ) {
+    const isMe = root === selfProcess.rank;
+    const rootProcess = getProcessByRank(root);
+    if (!isMe) {
+      await sliceAndSend(data, 0, sendCount, rootProcess, abort);
+    }
+
+    if (isMe) {
+      const writeToBufferByProcess = async (
+        process: Process,
+        received: unknown[],
+      ) => {
+        const { rank } = process;
+        const count = recvCounts[rank];
+        const offset = recvOffsets[rank];
+
+        for (let i = offset, j = 0; i < offset + count; i++, j++) {
+          buf[i] = received[j];
+        }
+      };
```

Comment:
todo(non-blocking): looks like this function can be extracted from method itself